Orange/widgets/utils/listfilter.py

```diff
@@ -0,0 +1,244 @@
+                rows = list(map(QModelIndex.row, selected))
+                for s1, s2 in reversed(list(slices(rows))):
+                    delslice(self.model(), s1, s2)
+            self.dragDropActionDidComplete.emit(res)
+
+    def dragEnterEvent(self, event):
+        """
+        Reimplemented from QListView.dragEnterEvent
+        """
+        if self.acceptsDropEvent(event):
+            event.accept()
+        else:
+            event.ignore()
+
+    def acceptsDropEvent(self, event):
+        """
+        Should the drop event be accepted?
+        """
+        # disallow drag/drops between windows
+        if event.source() is not None and \
+                event.source().window() is not self.window():
+            return False
+
+        mime = event.mimeData()
+        vars = mime.property('_items')
+        if vars is None:
+            return False
+
+        if not all(isinstance(var, self.__acceptedType) for var in vars):
+            return False
+
+        event.accept()
+        return True
+
+
+class VariableFilterProxyModel(QSortFilterProxyModel):
+    """ A proxy model for filtering a list of variables based on
+    their names and labels.
+
+    """
+    def __init__(self, parent=None):
+        super().__init__(parent)
+        self._filter_string = ""
+
+    def set_filter_string(self, filter):
+        self._filter_string = str(filter).lower()
+        self.invalidateFilter()
+
+    def filter_accepts_variable(self, var):
+        row_str = var.name + " ".join(("%s=%s" % item)
+                                      for item in var.attributes.items())
+        row_str = row_str.lower()
+        filters = self._filter_string.split()
+
+        return all(f in row_str for f in filters)
+
+    def filterAcceptsRow(self, source_row, source_parent):
+        model = self.sourceModel()
+        if isinstance(model, VariableListModel):
+            var = model[source_row]
+            return self.filter_accepts_variable(var)
+        else:
+            return True
+
+
+class CompleterNavigator(QObject):
+    """ An event filter to be installed on a QLineEdit, to enable
+    Key up/ down to navigate between posible completions.
+    """
+    def eventFilter(self, obj, event):
+        if (event.type() == QEvent.KeyPress and
+                isinstance(obj, QLineEdit)):
+            if event.key() == Qt.Key_Down:
+                diff = 1
+            elif event.key() == Qt.Key_Up:
+                diff = -1
+            else:
+                return False
+            completer = obj.completer()
+            if completer is not None and completer.completionCount() > 0:
+                current = completer.currentRow()
+                current += diff
+                completer.setCurrentRow(current % completer.completionCount())
+                completer.complete()
+            return True
+        else:
+            return False
+
+
+class VariablesFilter(QObject):
```

Comment:
I like the idea of this PR. Implementation is messy, though. It is a `QObject` (why?) that constructs a line edit, a view and a model, and stores all of them as attributes that are later extracted. What does this object represent?

If you take a closer look -- one uses it by constructing it, taking the three objects, and then discarding the object. This is then not a class but rather a function that returns two objects, line edit and a view. Model is accessible through a view or, even better, given as an argument to the function, which lets the widget construct a model it needs (`DomainModel`, `VariableListModel`...). The two update methods become enclosed functions, and the `add_*` methods are removed. All attributes (that are needed) become just local variables stored in closures.